src/fastfield/writer.rs

```diff
@@ -308,31 +295,61 @@ fn record_json_value_to_columnar_writer(
         return;
     }
     remaining_depth_limit -= 1;
+
     match json_val {
-        serde_json::Value::Null => {
-            // TODO handle null
-        }
-        serde_json::Value::Bool(bool_val) => {
-            columnar_writer.record_bool(doc, json_path_writer, *bool_val);
-        }
-        serde_json::Value::Number(json_number) => {
-            if let Some(numerical_value) = columnar_numerical_value(json_number) {
-                columnar_writer.record_numerical(doc, json_path_writer.as_str(), numerical_value);
-            }
-        }
-        serde_json::Value::String(text) => {
+        ReferenceValue::Null => {} // TODO: Handle null
+        ReferenceValue::Str(val) => {
             if let Some(text_analyzer) = tokenizer.as_mut() {
-                let mut token_stream = text_analyzer.token_stream(text);
+                let mut token_stream = text_analyzer.token_stream(val);
                 token_stream.process(&mut |token| {
                     columnar_writer.record_str(doc, json_path_writer.as_str(), &token.text);
                 })
             } else {
-                columnar_writer.record_str(doc, json_path_writer.as_str(), text);
+                columnar_writer.record_str(doc, json_path_writer.as_str(), val);
             }
         }
-        serde_json::Value::Array(arr) => {
-            for el in arr {
-                record_json_value_to_columnar_writer(
+        ReferenceValue::U64(val) => {
+            columnar_writer.record_numerical(
+                doc,
+                json_path_writer.as_str(),
+                NumericalValue::from(val),
+            );
+        }
+        ReferenceValue::I64(val) => {
+            columnar_writer.record_numerical(
+                doc,
+                json_path_writer.as_str(),
+                NumericalValue::from(val),
+            );
+        }
+        ReferenceValue::F64(val) => {
+            columnar_writer.record_numerical(
+                doc,
+                json_path_writer.as_str(),
+                NumericalValue::from(val),
+            );
+        }
+        ReferenceValue::Bool(val) => {
+            columnar_writer.record_bool(doc, json_path_writer, val);
+        }
+        ReferenceValue::Date(_) => {
+            unimplemented!("Datetime support in dynamic fields is not yet implemented")
```

Comment:
We can support that. There's a precision config missing on JSON, so we could just store it with the highest precision